dev/io.openliberty.org.crac.1.4/src/org/crac/LibertyCheckpoint.java

```diff
@@ -50,10 +50,19 @@ public void prepare() {
         callHooks(hooks.descendingIterator(), CheckpointHook::prepare, () -> new CheckpointException(Tr.formatMessage(tc, "CRAC_RESOURCE_CHECKPOINT_FAIL_CWWKC0551")));
     }
 
+    @Override
+    public void checkpointFailed() {
+        // CRaC says to call the afterRestore on each resource when checkpoint fails;
```

Comment:
When a `Resource` is registered with the `LibertyCheckpoint.register` method the code wraps the `Resource` in a Liberty `io.openliberty.checkpoint.spi.CheckpointHook` implementations where each `CheckpointHook` has a weak reference to the `Resource`.  This inner `CheckpointHook` then calls the appropriate `beforeCheckpoint` and `afterCheckpoint` methods on the wrapped resource when the `CheckpointHook` methods `prepare` and `restore` are called.

We then register this single `CheckpointHook` as an OSGi service that then calls out to the list of `CheckpointHook`s that wrap all the `Resource` objects registered.  Here the code is just calling out to the `CheckpointHook` wrapper objects to call the `CheckpointHook.restore` method and then that in turn calls the `afterRestore` method.

I think the confusion may be this extra level of indirection we have with the nested `CheckpointHook` wrappers.  It is not clear this wrapper is even needed.  I believe we could change the field `private static final Deque<CheckpointHook> hooks = new ConcurrentLinkedDeque<CheckpointHook>();` into a `Deque<WeakReference<Resource>` instead and avoid the wrappers completely.  This avoids the unnecessary `CheckpointHook` object and I think would make the code more clear.  Thanks for the observation about the confusion here.  I'll make some changes to get rid of this extra wrapper and indirection.

Reply:
It is a bit more complicated that I stated above.  Turns out we need some kind of "wrapper" for the following reasons:

1. We need to capture the `Context` used to register the `Resource` so we can call the `beforeCheckpoint(Context)` and `afterRestore(Context)` methods with the captured `Context` from registration.
2. We also need to hand exceptions thrown from the `Resource` methods so we can add them to a single exception as a suppressed exception.

This is all a bit more easy if we wrap the `Resource` objects in some other object.  A `CheckpointHook` is as good an object as any.  I don't plan to make any changes here for now.

Reply:
Thank you for the explanation on the CheckpointHooks objects Tom! This definitely helps my understanding of the placement of the hooks and how it references afterRestore since I couldn't make that connection at first. I just wanted to double check the working of this method to make sure I understood this approach.

Reply:
Ohh ok, sounds good. Then as you mentioned we'll go ahead with the existing changes, thank you Tom!